pkg/operator/rules.go

```diff
@@ -42,40 +43,62 @@ const (
 
 type PrometheusRuleSelector struct {
 	ruleFormat   RuleConfigurationFormat
+	version      semver.Version
 	ruleSelector labels.Selector
 	nsLabeler    *namespacelabeler.Labeler
 	ruleInformer *informers.ForResource
 	logger       log.Logger
 }
 
-func NewPrometheusRuleSelector(ruleFormat RuleConfigurationFormat, labelSelector *metav1.LabelSelector, nsLabeler *namespacelabeler.Labeler, ruleInformer *informers.ForResource, logger log.Logger) (*PrometheusRuleSelector, error) {
+func NewPrometheusRuleSelector(ruleFormat RuleConfigurationFormat, version string, labelSelector *metav1.LabelSelector, nsLabeler *namespacelabeler.Labeler, ruleInformer *informers.ForResource, logger log.Logger) (*PrometheusRuleSelector, error) {
+	componentVersion, err := semver.ParseTolerant(version)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse version")
+	}
+
 	ruleSelector, err := metav1.LabelSelectorAsSelector(labelSelector)
 	if err != nil {
 		return nil, errors.Wrap(err, "convert rule label selector to selector")
 	}
 
 	return &PrometheusRuleSelector{
 		ruleFormat:   ruleFormat,
+		version:      componentVersion,
 		ruleSelector: ruleSelector,
 		nsLabeler:    nsLabeler,
 		ruleInformer: ruleInformer,
 		logger:       logger,
 	}, nil
 }
 
-func generateRulesConfiguration(ruleformat RuleConfigurationFormat, promRule monitoringv1.PrometheusRuleSpec, logger log.Logger) (string, error) {
-	if ruleformat == PrometheusFormat {
-		// Unset partialResponseStrategy field.
-		for i := range promRule.Groups {
-			promRule.Groups[i].PartialResponseStrategy = ""
+func (pr *PrometheusRuleSelector) generateRulesConfiguration(promRule *monitoringv1.PrometheusRule) (string, error) {
+	logger := log.With(pr.logger, "prometheusrule", promRule.Name, "prometheusrule-namespace", promRule.Namespace)
+	promRuleSpec := promRule.Spec
+	minVersionLimits := semver.MustParse("2.31.0")
+	component := "Prometheus"
+
+	if pr.ruleFormat == ThanosFormat {
+		minVersionLimits = semver.MustParse("0.24.0")
+		component = "Thanos"
+	}
+
+	for i := range promRuleSpec.Groups {
+		if promRuleSpec.Groups[i].Limit != 0 {
+			if pr.version.LT(minVersionLimits) {
```

Comment:
(nit) a bit simpler to read.
```suggestion
		if promRuleSpec.Groups[i].Limit != 0 && pr.version.LT(minVersionLimits) {
```